proposed/2021/PackageNamespaces.md

```diff
@@ -0,0 +1,568 @@
+
+We also considered alternatives such as [`npm scopes`](https://docs.npmjs.com/cli/v7/using-npm/scope) which would include the package source & namespace on the package ID, but that did not fit the MSBuild expectations of our developer-base as the syntax was too foreign & potentially breaking.
+
+We then considered package grouping concepts which was proposed in [Central Version Package Management](https://github.com/NuGet/Home/issues/6764) which would allow control from the package declaration perspective compared to a source view. Although people found this promising as a concept, we found that without large adoption of the existing feature & limitation on `PackageReference` scenarios, that we would not be able to cover our hybrid ecosystem of .NET Framework & .NET Core users.
+
+Lastly, we considered package lock files which would ensure repeatability based on the contents of a package. For the sake of control to the user, this approach did not cover the many needs of limiting or filtering as there are currently no great ways to specify a package source nor package namespace within lock file tooling. We may reconsider this at a future date.
+
+## Prior Art
+
+<!-- What prior art, both good and bad are related to this proposal? -->
+<!-- Do other features exist in other ecosystems and what experience have their community had? -->
+<!-- What lessons from other communities can we learn from? -->
+<!-- Are there any resources that are relevent to this proposal? -->
+There's a number of features that exist in various ecosystems & layers that solve similar problems by providing the user more control of their software supply chain such as:
+
+- [npm (scopes)](https://docs.npmjs.com/about-scopes)
+- [Maven (banned dependencies)](https://maven.apache.org/enforcer/enforcer-rules/bannedDependencies.html)
+- [JFrog (exclude patterns)](https://jfrog.com/blog/yet-another-case-for-using-exclude-patterns-in-remote-repositories/)
+- [Azure DevOps (packaging exclusions)](https://docs.microsoft.com/en-us/dynamics365/fin-ops-core/dev-itpro/dev-tools/exclude-test-packages)
+- [Azure DevOps (.artifactIgnore)](https://docs.microsoft.com/en-us/azure/devops/artifacts/reference/artifactignore?view=azure-devops)
+- [GitHub Actions (exclude paths)](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#excluding-paths)
+
+## Unresolved Questions
+
+<!-- What parts of the proposal do you expect to resolve before this gets accepted? -->
+<!-- What parts of the proposal need to be resolved before the proposal is stabilized? -->
+<!-- What related issues would you consider out of scope for this proposal but can be addressed in the future? -->
+
+- Should strict namespaces mode be the default? That would deliver a `secure by default` namespacing experience.
```

Comment:
Yes.

If you're using this feature, I believe the general intention is to limit something to a single prefix. As such, I think the "default by secure" approach is a better choice.

Reply:
Seconded. At a minimum, nuget should reject restoring any package that's present on multiple feeds unless whitelisted by namespace and/or author.

Reply:
> nuget should reject restoring any package that's present on multiple feeds unless whitelisted

I love this. Very easy threat fix. I assume whitelisting feeds (like private beta vs prod) not just packages.

Reply:
or nuget should implement the equivalent of @scope/package like npm
and assign a feed to an @scope
default @scope could be associated to the user or org pushing the package for backward compat and enrich existing package so there's no breaking change

that would address what's mentionned above and at the same time facilitate multi feed

on top of that nuget v7 could make the use of scope mandatory

which would also fix the "name reservation" with something that looks way more clear especially even doing mono repo of multi library